ERCS/erc-7588.md

```diff
@@ -0,0 +1,152 @@
+            "description": "Describes the MIME type of the blobs. The MIME type should be defined in RFC 2046 (https://www.rfc-editor.org/rfc/rfc2046)"
+        },
+        "extras": {
+            "type": "string",
+            "description": "Dynamic extra information related to the blobs"
+        },
+        "blobs": {
+            "type": "array",
+            "description": "Metadata of the i'th blob. This is optional and overlays the upper level properties if provided",
+            "items": {
+                "description": {
+                    "type": "string",
+                    "description": "Describes the content of the i'th blob"
+                },
+                "content_type": {
+                    "type": "string",
+                    "description": "Describes the MIME type of the i'th blob. The MIME type should be defined in RFC 2046 (https://www.rfc-editor.org/rfc/rfc2046)"
+                },
+                "extras": {
+                    "type": "string",
+                    "description": "Dynamic extra information related to the i'th blob"
+                },
+            }
+        }
+    }
+}
+```
+
+For example, suppose Vitalik wants to send a blob transaction carrying two blobs explaining “data availability sampling.” He could include a paragraph of textual explanation in the first blob and an illustration image in the second blob. The corresponding metadata JSON object would look like this:
+
+```json
+{
+    "originator": "Vitalik Buterin",
+    "description": "An illustration of data availability sampling",
+    "blobs": [
+      {
+        "content_type": "text/plain",
+        "description": "This blob contains a description text of the illustration. It's a hex encoded UTF-8 string."
+      },
+      {
+        "content_type": "image/png",
+        "description": "This blob contains the illustration image data in base64 format. It's a hex encoded RFC 2397 (https://www.rfc-editor.org/rfc/rfc2397) data URL."
+      },
+    ]
+  }
+```
+
+The complete blob transaction would include this metadata in the data field, along with other relevant fields:
+
+```json
+{
+  "blobVersionedHashes": ["0x...", "0x..."],
+  "chainId": 11155111, // Supposing the blob transaction is posted to Sepolia
+  "type": "eip4844",
+  "to": "0x0000000000000000000000000000000000000000",
+  "gas": 28236,
+  "data": "0x..", // Hex value of the string representation of the above metadata JSON object
+  "nonce": 18,
+  "maxFeePerBlobGas": 1073677089,
+  "maxFeePerGas": 1213388073,
+  "maxPriorityFeePerGas": 1165808679,
+  "sidecars": [
+    { "blob": "0x...", "commitment": "0x...", "proof": "0x..." },
+    { "blob": "0x...", "commitment": "0x...", "proof": "0x..." }
+  ]
+}
+```
+
+### Metadata Encoding
+
+The metadata JSON object should be converted to a string representation and encoded to a hexadecimal value. For instance, an empty metadata JSON object should be converted to "{}" and encoded as "0x7b7d".
```

Comment:
This is still a bit weird, and now contradictory. You're telling readers to encode the JSON to hexadecimal. That step is necessary in languages like JavaScript where byte arrays aren't as common. It makes no sense for languages like Rust where you need to pass in a byte array; it's just confusing.

I _think_ you're trying to say "set the calldata of the transaction to the string representation of the metadata JSON object, encoded in UTF-8"?

Reply:
Hi Sam, thanks for helping clarifying this. I've updated the PR to rephrase it as below, hopefully making it more clearer:

> The blob transaction's calldata (i.e., the data field) should be set to the hexadecimal string representation of the metadata JSON object, encoded in UTF-8. For instance, calldata "0x7b7d" indicates an empty metadata JSON object "{}".

BTW, I've also authored a script to demonstrate how to send a blob transaction with this metadata extension, which you could check if interested:

https://github.com/gavfu/exp-blobtx/blob/main/src/app.ts

Reply:
I feel like we're still talking past each other. There is _no_ hexadecimal involved in this standard, unless I'm missing something.

Take the following Rust code:

```rust
use ethers::{
    core::{types::TransactionRequest, utils::Anvil},
    providers::{Http, Middleware, Provider},
};
use eyre::Result;
use serde_json::json;
use std::convert::TryFrom;

#[tokio::main]
async fn main() -> Result<()> {
    let anvil = Anvil::new().spawn();

    // connect to the network
    let provider = Provider::<Http>::try_from(anvil.endpoint())?;
    let accounts = provider.get_accounts().await?;
    let from = accounts[0];
    let to = accounts[1];

    let metadata = json!({});
    let metadata_string = serde_json::to_string(&metadata)?;
    println!("{metadata_string}"); // Prints "{}"

    let metadata_bytes = metadata_string.into_bytes();
    println!("{metadata_bytes:?}"); // Prints "[123, 125]"

    // craft the tx
    let tx = TransactionRequest::new()
        .to(to)
        .data(metadata_bytes)
        .from(from);

    provider.send_transaction(tx, None).await?.await?;

    Ok(())
}
```

There is absolutely no hexadecimal involved. The data of the transaction is set to the UTF-8 encoding of `metadata_string`. Saying that you're encoding something as hexadecimal is confusing because that's a JavaScript/TypeScript oddity, not something that's generally applicable to your standard.

Reply:
Hi Sam, really appreciate your detailed explanation. Now I understand your points, and updated the draft accordingly. Please let me know if you have any other comments or suggestions.  :)